src/cryptsetup/cryptsetup.c

```diff
@@ -2416,20 +2393,52 @@ static int run(int argc, char *argv[]) {
                 }
 #endif
 
-                bool use_cached_passphrase = true;
+                bool use_cached_passphrase = true, try_discover_key = !key_file;
                 _cleanup_strv_free_erase_ char **passwords = NULL;
                 for (tries = 0; arg_tries == 0 || tries < arg_tries; tries++) {
+                        _cleanup_(erase_and_freep) void *key_data = NULL;
+                        size_t key_data_size;
+
                         log_debug("Beginning attempt %u to unlock.", tries);
 
                         /* When we were able to acquire multiple keys, let's always process them in this order:
                          *
                          *    1. A key acquired via PKCS#11 or FIDO2 token, or TPM2 chip
-                         *    2. The discovered key: i.e. key_data + key_data_size
-                         *    3. The configured key: i.e. key_file + arg_keyfile_offset + arg_keyfile_size
-                         *    4. The empty password, in case arg_try_empty_password is set
-                         *    5. We enquire the user for a password
+                         *    2. The configured or discovered key, of which both are exclusive and optional
+                         *    3. The empty password, in case arg_try_empty_password is set
+                         *    4. We enquire the user for a password
                          */
 
+                        if (try_discover_key) {
+                                _cleanup_free_ char *bindname = NULL;
+                                const char *fn, *client = "cryptsetup";
```

Comment:
maybe don't initialize `client` here, but instead move it to a final `else` branch in the expression below.